In [1]:
#Standardizing only the Numerical Variables(Creating a custom Scaler)

#Import libraries
import pandas as pd
import numpy as np

In [2]:
#Load the data
data_preprocessed = pd.read_csv('Absenteeism_preprocessed.csv')

In [3]:
data_preprocessed

reason_1  reason_2  reason_3  reason_4  Transportation Expense  \
0           0         0         0         1                     289   
1           0         0         0         0                     118   
2           0         0         0         1                     179   
3           1         0         0         0                     279   
4           0         0         0         1                     289   
5           0         0         0         1                     179   
6           0         0         0         1                     361   
7           0         0         0         1                     260   
8           0         0         1         0                     155   
9           0         0         0         1                     235   
10          1         0         0         0                     260   
11          1         0         0         0                     260   
12          1         0         0         0                     260   
13          1         0         0         0                     179   
14          0         0         0         1                     179   
15          1         0         0         0                     246   
16          0         0         0         1                     179   
17          0         0         1         0                     179   
18          1         0         0         0                     189   
19          0         0         0         1                     248   
20          1         0         0         0                     330   
21          1         0         0         0                     179   
22          1         0         0         0                     361   
23          0         0         0         1                     260   
24          0         0         1         0                     289   
25          0         0         0         1                     361   
26          0         0         0         1                     289   
27          0         0         0         1                     157   
28          0         0         1         0                     289   
29          0         0         0         1                     179   
..        ...       ...       ...       ...                     ...   
670         0         0         0         1                     155   
671         0         0         1         0                     225   
672         1         0         0         0                     118   
673         0         0         0         1                     179   
674         0         0         0         1                     235   
675         0         0         1         0                     225   
676         0         0         0         1                     235   
677         1         0         0         0                     118   
678         0         0         0         1                     235   
679         1         0         0         0                     179   
680         0         0         0         1                     291   
681         1         0         0         0                     225   
682         0         0         1         0                     300   
683         0         0         0         1                     225   
684         0         0         0         1                     179   
685         0         0         0         1                     118   
686         1         0         0         0                     118   
687         0         0         0         1                     118   
688         0         0         0         0                     118   
689         0         0         0         1                     179   
690         0         0         0         0                     378   
691         0         1         0         0                     179   
692         1         0         0         0                     155   
693         1         0         0         0                     235   
694         0         0      

In [4]:
#Targets
data_preprocessed['Absenteeism Time in Hours'].median()

3.0

In [5]:
targets = np.where(data_preprocessed['Absenteeism Time in Hours']>data_preprocessed['Absenteeism Time in Hours'].median(), 1, 0)

In [6]:
data_preprocessed['Excessive Absenteeism'] = targets

In [7]:
data_preprocessed.head(10)

reason_1  reason_2  reason_3  reason_4  Transportation Expense  \
0         0         0         0         1                     289   
1         0         0         0         0                     118   
2         0         0         0         1                     179   
3         1         0         0         0                     279   
4         0         0         0         1                     289   
5         0         0         0         1                     179   
6         0         0         0         1                     361   
7         0         0         0         1                     260   
8         0         0         1         0                     155   
9         0         0         0         1                     235   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          0   
1                13   50                  239.554               31          0   
2                51   38                  239.554               31          0   
3                 5   39                  239.554               24          0   
4                36   33                  239.554               30          0   
5                51   38                  239.554               31          0   
6                52   28                  239.554               27          0   
7                50   36                  239.554               23          0   
8                12   34                  239.554               25          0   
9                11   37                  239.554               29          1   

   Children  Pets  Month Value  Day of the Week  Absenteeism Time in Hours  \
0         2     1            7                1                          4   
1         1     0            7                1                          0   
2         0     0            7                2                          2   
3         2     0            7                3                          4   
4         2     1            7                3                          2   
5         0     0            7                4                          2   
6         1     4            7                4                          8   
7         4     0            7                4                          4   
8         2     0            7                0                         40   
9         1     1            7                0                          8   

   Excessive Absenteeism  
0                      1  
1                      0  
2                      0  
3                      1  
4                      0  
5                      0  
6                      1  
7                      1  
8                      1  
9                      1

In [8]:
targets.sum()/targets.shape[0]

0.45571428571428574

In [9]:
data_with_targets = data_preprocessed.drop(['Absenteeism Time in Hours'], axis=1)

In [10]:
data_with_targets is data_preprocessed

False

In [11]:
data_with_targets.head()

reason_1  reason_2  reason_3  reason_4  Transportation Expense  \
0         0         0         0         1                     289   
1         0         0         0         0                     118   
2         0         0         0         1                     179   
3         1         0         0         0                     279   
4         0         0         0         1                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          0   
1                13   50                  239.554               31          0   
2                51   38                  239.554               31          0   
3                 5   39                  239.554               24          0   
4                36   33                  239.554               30          0   

   Children  Pets  Month Value  Day of the Week  Excessive Absenteeism  
0         2     1            7                1                      1  
1         1     0            7                1                      0  
2         0     0            7                2                      0  
3         2     0            7                3                      1  
4         2     1            7                3                      0

In [12]:
#Selecting the inputs for the logistic regression
data_with_targets.shape

(700, 15)

In [13]:
unscaled_inputs = data_with_targets.iloc[:, :-1]
unscaled_inputs

reason_1  reason_2  reason_3  reason_4  Transportation Expense  \
0           0         0         0         1                     289   
1           0         0         0         0                     118   
2           0         0         0         1                     179   
3           1         0         0         0                     279   
4           0         0         0         1                     289   
5           0         0         0         1                     179   
6           0         0         0         1                     361   
7           0         0         0         1                     260   
8           0         0         1         0                     155   
9           0         0         0         1                     235   
10          1         0         0         0                     260   
11          1         0         0         0                     260   
12          1         0         0         0                     260   
13          1         0         0         0                     179   
14          0         0         0         1                     179   
15          1         0         0         0                     246   
16          0         0         0         1                     179   
17          0         0         1         0                     179   
18          1         0         0         0                     189   
19          0         0         0         1                     248   
20          1         0         0         0                     330   
21          1         0         0         0                     179   
22          1         0         0         0                     361   
23          0         0         0         1                     260   
24          0         0         1         0                     289   
25          0         0         0         1                     361   
26          0         0         0         1                     289   
27          0         0         0         1                     157   
28          0         0         1         0                     289   
29          0         0         0         1                     179   
..        ...       ...       ...       ...                     ...   
670         0         0         0         1                     155   
671         0         0         1         0                     225   
672         1         0         0         0                     118   
673         0         0         0         1                     179   
674         0         0         0         1                     235   
675         0         0         1         0                     225   
676         0         0         0         1                     235   
677         1         0         0         0                     118   
678         0         0         0         1                     235   
679         1         0         0         0                     179   
680         0         0         0         1                     291   
681         1         0         0         0                     225   
682         0         0         1         0                     300   
683         0         0         0         1                     225   
684         0         0         0         1                     179   
685         0         0         0         1                     118   
686         1         0         0         0                     118   
687         0         0         0         1                     118   
688         0         0         0         0                     118   
689         0         0         0         1                     179   
690         0         0         0         0                     378   
691         0         1         0         0                     179   
692         1         0         0         0                     155   
693         1         0         0         0                     235   
694         0         0      

In [14]:
#Standardizing the data
#from sklearn.preprocessing import StandardScaler
#absenteeism_scaler = StandardScaler()

In [15]:
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import StandardScaler

class CustomScaler(BaseEstimator, TransformerMixin):
    def __init__(self, columns, copy=True, with_mean=True, with_std=True):
        self.scaler = StandardScaler(copy, with_mean, with_std)
        self.columns = columns
        self.mean_ = None
        self.var_ = None
        
    def fit(self, X, y=None):
        self.scaler.fit(X[self.columns], y)
        self.mean_ = np.mean(X[self.columns])
        self.var_ = np.var(X[self.columns])
        return self
    
    def transform(self, X, y=None, copy=None):
        init_col_order = X.columns
        X_scaled = pd.DataFrame(self.scaler.transform(X[self.columns]), columns=self.columns)
        X_not_scaled = X.loc[:,~X.columns.isin(self.columns)]
        return pd.concat([X_not_scaled, X_scaled], axis=1)[init_col_order]

In [16]:
unscaled_inputs.columns.values

array(['reason_1', 'reason_2', 'reason_3', 'reason_4',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'Month Value', 'Day of the Week'], dtype=object)

In [17]:
columns_to_scale = ['Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index','Children', 
                    'Pets', 'Month Value', 'Day of the Week']

In [18]:
absenteeism_scaler = CustomScaler(columns_to_scale)

In [19]:
absenteeism_scaler.fit(unscaled_inputs)

CustomScaler(columns=['Transportation Expense', 'Distance to Work', 'Age',
                      'Daily Work Load Average', 'Body Mass Index', 'Children',
                      'Pets', 'Month Value', 'Day of the Week'],
             copy=None, with_mean=None, with_std=None)

In [20]:
scaled_inputs = absenteeism_scaler.transform(unscaled_inputs)

In [21]:
scaled_inputs

reason_1  reason_2  reason_3  reason_4  Transportation Expense  \
0           0         0         0         1                1.005844   
1           0         0         0         0               -1.574681   
2           0         0         0         1               -0.654143   
3           1         0         0         0                0.854936   
4           0         0         0         1                1.005844   
5           0         0         0         1               -0.654143   
6           0         0         0         1                2.092381   
7           0         0         0         1                0.568211   
8           0         0         1         0               -1.016322   
9           0         0         0         1                0.190942   
10          1         0         0         0                0.568211   
11          1         0         0         0                0.568211   
12          1         0         0         0                0.568211   
13          1         0         0         0               -0.654143   
14          0         0         0         1               -0.654143   
15          1         0         0         0                0.356940   
16          0         0         0         1               -0.654143   
17          0         0         1         0               -0.654143   
18          1         0         0         0               -0.503235   
19          0         0         0         1                0.387122   
20          1         0         0         0                1.624567   
21          1         0         0         0               -0.654143   
22          1         0         0         0                2.092381   
23          0         0         0         1                0.568211   
24          0         0         1         0                1.005844   
25          0         0         0         1                2.092381   
26          0         0         0         1                1.005844   
27          0         0         0         1               -0.986140   
28          0         0         1         0                1.005844   
29          0         0         0         1               -0.654143   
..        ...       ...       ...       ...                     ...   
670         0         0         0         1               -1.016322   
671         0         0         1         0                0.040034   
672         1         0         0         0               -1.574681   
673         0         0         0         1               -0.654143   
674         0         0         0         1                0.190942   
675         0         0         1         0                0.040034   
676         0         0         0         1                0.190942   
677         1         0         0         0               -1.574681   
678         0         0         0         1                0.190942   
679         1         0         0         0               -0.654143   
680         0         0         0         1                1.036026   
681         1         0         0         0                0.040034   
682         0         0         1         0                1.171843   
683         0         0         0         1                0.040034   
684         0         0         0         1               -0.654143   
685         0         0         0         1               -1.574681   
686         1         0         0         0               -1.574681   
687         0         0         0         1               -1.574681   
688         0         0         0         0               -1.574681   
689         0         0         0         1               -0.654143   
690         0         0         0         0                2.348925   
691         0         1         0         0               -0.654143   
692         1         0         0         0               -1.016322   
693         1         0         0         0                0.190942   
694         0         0      

In [22]:
#Splitting the date for training and testing
from sklearn.model_selection import train_test_split

In [23]:
train_test_split(scaled_inputs, targets)

[     reason_1  reason_2  reason_3  reason_4  Transportation Expense  \
 686         1         0         0         0               -1.574681   
 549         0         0         0         0                1.005844   
 391         0         0         1         0                0.356940   
 350         0         0         0         1                0.190942   
 599         0         0         0         1                0.040034   
 555         1         0         0         0               -0.654143   
 307         0         0         0         1                1.036026   
 327         1         0         0         0               -1.574681   
 364         1         0         0         0                1.036026   
 197         0         0         1         0                0.568211   
 372         0         0         0         1                1.005844   
 443         0         0         0         1               -1.574681   
 146         0         0         0         1                0.04

In [24]:
x_train, x_test, y_train, y_test = train_test_split(scaled_inputs, targets)
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)

(525, 14) (525,)
(175, 14) (175,)


In [25]:
x_train, x_test, y_train, y_test = train_test_split(scaled_inputs, targets, train_size=0.8, shuffle=True, random_state=20)

In [26]:
y_test.shape

(140,)

In [27]:
#Fitting the Model and Assessing it's accuracy
from sklearn.linear_model import LogisticRegression
from sklearn import metrics

In [28]:
##training the model
reg = LogisticRegression()

In [29]:
reg.fit(x_train, y_train)

//anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [30]:
reg.score(x_train, y_train)

0.7714285714285715

In [31]:
##Manually check the accuracy
model_outputs = reg.predict(x_train)
model_outputs

array([0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0,

In [32]:
np.sum(model_outputs == y_train)/model_outputs.shape[0]  #the accuracy from the model is correct!

0.7714285714285715

In [33]:
#Creating a Summary Table with the coefficient and Intercept
reg.intercept_

array([-1.47082548])

In [34]:
reg.coef_

array([[ 2.62614283,  0.84432   ,  2.93971875,  0.67771914,  0.60695978,
        -0.00741802, -0.16862442, -0.00383568,  0.26749958, -0.23085763,
         0.35572764, -0.28559431,  0.16145652, -0.08257256]])

In [35]:
feature_name = unscaled_inputs.columns.values

In [36]:
summary_table = pd.DataFrame(columns=['Feature name'], data=feature_name)
summary_table['Coefficient'] = np.transpose(reg.coef_)
summary_table

Feature name  Coefficient
0                  reason_1     2.626143
1                  reason_2     0.844320
2                  reason_3     2.939719
3                  reason_4     0.677719
4    Transportation Expense     0.606960
5          Distance to Work    -0.007418
6                       Age    -0.168624
7   Daily Work Load Average    -0.003836
8           Body Mass Index     0.267500
9                 Education    -0.230858
10                 Children     0.355728
11                     Pets    -0.285594
12              Month Value     0.161457
13          Day of the Week    -0.082573

In [37]:
##Adding interception at the first place
summary_table.index = summary_table.index + 1
summary_table.loc[0] = ['Intercept', reg.intercept_[0]]
summary_table = summary_table.sort_index()
summary_table

Feature name  Coefficient
0                 Intercept    -1.470825
1                  reason_1     2.626143
2                  reason_2     0.844320
3                  reason_3     2.939719
4                  reason_4     0.677719
5    Transportation Expense     0.606960
6          Distance to Work    -0.007418
7                       Age    -0.168624
8   Daily Work Load Average    -0.003836
9           Body Mass Index     0.267500
10                Education    -0.230858
11                 Children     0.355728
12                     Pets    -0.285594
13              Month Value     0.161457
14          Day of the Week    -0.082573

In [38]:
#Interpreting the coefficients for our Problem
summary_table['odds_ratio'] = np.exp(summary_table.Coefficient)
summary_table

Feature name  Coefficient  odds_ratio
0                 Intercept    -1.470825    0.229736
1                  reason_1     2.626143   13.820360
2                  reason_2     0.844320    2.326395
3                  reason_3     2.939719   18.910527
4                  reason_4     0.677719    1.969381
5    Transportation Expense     0.606960    1.834845
6          Distance to Work    -0.007418    0.992609
7                       Age    -0.168624    0.844826
8   Daily Work Load Average    -0.003836    0.996172
9           Body Mass Index     0.267500    1.306693
10                Education    -0.230858    0.793852
11                 Children     0.355728    1.427219
12                     Pets    -0.285594    0.751567
13              Month Value     0.161457    1.175221
14          Day of the Week    -0.082573    0.920745

In [39]:
summary_table.sort_values('odds_ratio', ascending=False)

Feature name  Coefficient  odds_ratio
3                  reason_3     2.939719   18.910527
1                  reason_1     2.626143   13.820360
2                  reason_2     0.844320    2.326395
4                  reason_4     0.677719    1.969381
5    Transportation Expense     0.606960    1.834845
11                 Children     0.355728    1.427219
9           Body Mass Index     0.267500    1.306693
13              Month Value     0.161457    1.175221
8   Daily Work Load Average    -0.003836    0.996172
6          Distance to Work    -0.007418    0.992609
14          Day of the Week    -0.082573    0.920745
7                       Age    -0.168624    0.844826
10                Education    -0.230858    0.793852
12                     Pets    -0.285594    0.751567
0                 Intercept    -1.470825    0.229736

#Interpreting the coefficients of the Logistic Regression
1. important values 3,1,2,4,5,11,9 & 10 Education,12 Pets
2. unimportant variables: 13,8,6,14
3. 5 reasons variable, 0-no reason(baseline model-when no reason is given), 1-various diseases, 2-pregnacy and giving birth, 3-poisioning, 4-light diseases
*poisoning: The weight means the odds of someone being excessively absent after being poisoned are 20 times higher than when no reason was reported.
*various diseases: the normal absenteeism case, you got sick, you skipped work. A person who has reported this is 14 times more likely to be excessively absent than a person who didn't specify a reason
*Pregnacy and giving birth: pregnacy check and come back to work, from time to time, there's some emergencies, but the odds ratio is only 2.3 times more likely to be excessively absent than the base model.

4. Transportation Expense: is one of our standardized variables, we don't have direct interpretability of it. the odds ratio implies that for 1 standardized unit or for 1 standard deviation increase in transportation expense, it is close to twice as likely to be excessively absent
5. Pet: for each additional standardized unit of pet, the odds are 1-0.751567 = 25% lower than the base model(no pet) explanation: maybe if you have many pets, you are probably not taking care of them on your own, you will have someone to take pets to the doctor if something is wrong.
6. intercept: no meaning, in ML, you can say the intercept/Bias 'calibrates' the model.
